### PR TITLE
MTLModel+Difference: handle nil and incompatible MTLModel argument.

### PR DESCRIPTION
**Pull request by arangato**
Raising from differenceFrom has negative effect of disrupting tests and
makes getting to the mismatch harder. Therefore returning a descriptive
message for each edge case.

### CODE REVIEW

File: MantleTests/MTLModelDifferenceSpec.m [hunk 1]
```diff
@@ -10,10 +10,32 @@
 
 QuickSpecBegin(MTLModelDifferenceSpec)
 
-it(@"should raise on mismatching argument type", ^{
+it(@"should return message about expected type", ^{
   MTLTestModel *model = [[MTLTestModel alloc] init];
   NSAssert(model, @"model must be non nil");
-  expect([model differenceFrom:@1]).to(raiseException());
+  MTLTestModel *nilModel = nil;
+  NSArray *difference = [model differenceFrom:nilModel];
```
Comment (byohay) [inline MantleTests/MTLModelDifferenceSpec.m]: Just pass `differenceFrom:nil`

File: Mantle/MTLModel+Difference.m [hunk 2]
```diff
@@ -10,11 +10,29 @@ @implementation MTLModel (Difference)
...
+  if (!actual) {
+    return @[
+      [NSString stringWithFormat:@"Expected: object of type %@, got: nil", NSStringFromClass(self.class)]
+    ];
+  }
 
   NSMutableArray<NSString *> *differences = [NSMutableArray array];
+  if (![actual isKindOfClass:self.class]) {
+    [differences addObject:[NSString stringWithFormat:@"Expected: object of type %@, got: %@",
+                            NSStringFromClass(self.class), NSStringFromClass([object class])]];
```
Comment (byohay) [inline Mantle/MTLModel+Difference.m]: `class` is a method, use `[self class]`. Same for other places.

File: Mantle/MTLModel+Difference.m [hunk 3]
```diff
@@ -10,11 +10,29 @@ @implementation MTLModel (Difference)
...
-  NSAssert2([object isKindOfClass:self.class], @"Expected object of type %@, got : %@",
-            self.class, [object class]);
+  if (!actual) {
+    return @[
+      [NSString stringWithFormat:@"Expected: object of type %@, got: nil", NSStringFromClass(self.class)]
+    ];
+  }
 
   NSMutableArray<NSString *> *differences = [NSMutableArray array];
+  if (![actual isKindOfClass:self.class]) {
```
Comment (byohay) [inline Mantle/MTLModel+Difference.m]: You probably want to know if it's exactly the same class, not just inherits from it (Use `isEqual:`). Add a test for this.

File: Mantle/MTLModel+Difference.m [hunk 4]
```diff
@@ -10,11 +10,29 @@ @implementation MTLModel (Difference)
...
+                            NSStringFromClass(self.class), NSStringFromClass([object class])]];
+    if (![object isKindOfClass:MTLModel.class]) {
+      return differences;
+    }
+  }
+
   for (NSString *key in self.class.propertyKeys) {
+    if (![actual.class.propertyKeys containsObject:key]) {
+      return @[
+        [NSString stringWithFormat:@"Expected: object of type %@, got: %@",
```
Comment (byohay) [inline Mantle/MTLModel+Difference.m]: I think you need to either:
- Do a full diff between the models, even if they don't have the same property keys. Print a message for every key that's found in one but not the other.
- Don't handle models that aren't exactly of the same type.

I prefer the second one, I think that comparing between objects that don't have the same type doesn't have a lot of value, all you need is to know they don't have the same type.

Anyway if you go with the first, instead of going over `self.class.propertyKeys`, you need to take the union of `propertyKeys` from both of them.

**Review by byohay — CHANGES_REQUESTED** [review]
round.

File: MantleTests/MTLModelDifferenceSpec.m (hunk 1 above)
Comment (arangato) [reply MantleTests/MTLModelDifferenceSpec.m]: No can do. Xcode complains I'm passing `nil` to a method with non-nil argument.

File: Mantle/MTLModel+Difference.m (hunk 3 above)
Comment (arangato) [reply Mantle/MTLModel+Difference.m]: Originally, I did intended to support difference between related classes, but I'm getting more and more discouraged as the time passes. So I'll use class equality as you suggested and we'll wait for the first real need for something more elaborate.

File: Mantle/MTLModel+Difference.m (hunk 4 above)
Comment (arangato) [reply Mantle/MTLModel+Difference.m]: dropping it. same class only.

**Comment by arangato** [comment]
Round done.

File: Mantle/MTLModel+Difference.m [hunk 5]
```diff
@@ -7,11 +7,21 @@
...
 
-- (NSArray<NSString *> *)differenceFrom:(id)object {
+- (NSArray<NSString *> *)differenceFrom:(nullable id)object {
   MTLModel *actual = (MTLModel *)object;
 
-  NSAssert2([object isKindOfClass:self.class], @"Expected object of type %@, got : %@",
-            self.class, [object class]);
+  if (!actual) {
+    return @[
+      [NSString stringWithFormat:@"Expected: object of type %@, got: nil", NSStringFromClass(self.class)]
```
Comment (byohay) [inline Mantle/MTLModel+Difference.m]: `[self class]`

**Review by byohay — CHANGES_REQUESTED** [review]
F&S&M.

**Comment by arangato** [comment]
Fixed and squashed.

**Review by byohay — APPROVED** [review]
